File: mckp-19/src/main/java/problem/OneOrNoneForClass.java
package problem;

import model.ComputeSolutionImp;

/**
 * 
 */
public class OneOrNoneForClass extends ComputeSolutionImp {

    @Override
    public boolean checkClass(final int items, final int last, final int[] classes) {
        return classes[items] == last;
    }

    @Override
    public boolean compute(final int items, final int object, final int q, final int[] classes, final int[][] matrix) {
        int maxProfitForEachItem = 0;
        for (int k = 1; k <= items; k++) {
            if (classes[k] == classes[object] && matrix[q][k] > maxProfitForEachItem) {
                maxProfitForEachItem = matrix[q][k];
            }
        }
        return matrix[q][object] == maxProfitForEachItem;
    }

    @Override
    public boolean getSolution(final int profitLast, final int profitCurrent) {
        return profitCurrent > profitLast;
    }

    @Override
    public int getProfit(final int items, final int itemClass, final int[] knapsackCapacity, final int[] classes) {
        int maxProfit = 0;
        for (int k = 1; k < items; k++) {
            if(classes[k] == itemClass && knapsackCapacity[k] > maxProfit) {
                maxProfit = knapsackCapacity[k];
            }
        }
        return maxProfit;
    }
}

File: mckp-19/c++/OneOrNoneForClass.cpp
/**
 * This class permit the choice of an element for each class. Then it deals 
 * with implementing and managing the variant of the MCKP in which one can 
 * choose or not an element by class.
 */

#include "OneOrNoneForClass.h"

using namespace std;

namespace problem
{
	using ComputeSolutionImp = model::ComputeSolutionImp;

	bool OneOrNoneForClass::checkClass(int const items, int const last, vector<int> &classes)
	{
		return classes[items] == last;
	}

	bool OneOrNoneForClass::compute(int const items, int const object, int const q, vector<int> &classes, vector<vector<int>> &matrix)
	{
		int maxProfitForEachItem = 0;
		for (int k = 1; k <= items; k++)
		{
			if (classes[k] == classes[object] && matrix[q][k] > maxProfitForEachItem)
			{
				maxProfitForEachItem = matrix[q][k];
			}
		}
		return matrix[q][object] == maxProfitForEachItem;
	}

	bool OneOrNoneForClass::getSolution(int const profitLast, int const profitCurrent)
	{
		return profitCurrent > profitLast;
	}

	int OneOrNoneForClass::getProfit(int const items, int const itemClass, vector<int> &knapsackCapacity, vector<int> &classes)
	{
		int maxProfit = 0;
		for (int k = 1; k < items; k++)
		{
			if (classes[k] == itemClass && knapsackCapacity[k] > maxProfit)
			{
				maxProfit = knapsackCapacity[k];
			}
		}
		return maxProfit;
	}
}
